Ticket 776 orientation

pkienzle opened the pull request:
Use model-centric representation of jitter for angular dispersity, with psi representing rotation about the polar axis (length c in parallelepiped), phi about the equatorial major axis (length b) and theta about the equatorial minor axis (length a).  Viewing angle is relative to the beam, with psi rotating about the long axis, theta rotating into the xy-plane, then phi rotating within the xy-plane.

Use ```explore/jitter.py [MODEL]``` where MODEL is one of sphere, cylinder, ellipsoid, parallelepiped, triaxial_ellipsoid, bcc_paracrystal to interactively explore the relationship between viewing angle (theta, phi, psi) and jitter (dtheta, dphi, dpsi).  Default model is parallelepiped.  It shows the dispersity mesh over theta-phi where the I(q) will be evaluated.

Use ```explore/check1d.py MODEL -random``` to compare the 1-d model with the 2-d model using theta jitter ranging over [-pi, pi] and phi jitter ranging over [-2 pi, 2 pi].  These curves should match.  Use ```sascomp -sphere[=nsteps] model``` to show the full 2-d pattern with jitter on the full sphere. 

The angle transformations are generated by ```expore/angles.py``` from the symbolic math package sympy.  To use a different definition of angles, update the rotation matrices and rerun the code to generate the new angle transformation code.  The resulting code should be pasted into ```sasmodels/kernel_iq.c```.

Magnetism code was updated to make it simpler.  In the process, I discovered that cross-section weight was being applied to the effective sld, not the computed I(q) for the effective sld.  I changed this but did not cross check against sasview 3 or 4.0.  The magnetism calculations need to be reviewed to make sure the angle definitions are consistent with the updated interpretation of view and jitter.  Test values for magnetic scattering should be added.

Problem 1: bcc and fcc tests are not passing.  Likely we have changed the definition of angles relative to sasview and have not updated the tests to the appropriate qx, qy.

Problem 2: the 2-d integration test is not working correctly for asymmetric patterns at low q.  Use ```explore/asymint.py``` to show the results of the integral as computed by various integration schemes with different numbers of points.  The existing 1D kernels appear to give the correct result.

The first problem needs to be addressed before pulling.

I will continue to looking into the second problem, but it should not hold up this pull request.  This algorithm is much less wrong than the existing algorithm, particularly for smaller angular dispersity.

RichardHeenan commented:
Some initial testing ...
explore/jitter.py has a,b,c = 10,40,100   could we optionally add parameters to the command line please.
Oddly my version and Dirk's were not the same, I had to comment out ax.axis('square'), possibly a matplotlib version issue? I also get some error messages on closing the plot.

Trying out explore/check1d on some models (-double! openMP as I don't have openCL on laptop), so far ...
**sphere** - theta_pd does not exist
**cylinder** - OK
**ellipsoid** - mostly OK, some issues at high Q when very asymmetric shape.
**triaxial_ellipsoid** - sphere like is OK, issues at high Q when more prolate or oblate e.g. 
explore/check1d.py triaxial_ellipsoid radius_polar=40 radius_equat_minor=50 radius_equat_major=80 -double!
 which I would expect to be well behaved, needs some investigation?
**parallelepiped** - some issues at hgh Q when large but not huge, e.g.
explore/check1d.py parallelepiped length_a=100 length_b=120 length_c=150 -double!
**bcc_paracrystal** - is a mess as Paul said above, will have a look later.

RichardHeenan commented:
oopps I hit the wrong button, did not mean to close this, sorry

pkienzle commented:
Added size option to command line.  For example, ```explore/jitter.py ellipsoid 40,40,4000``` would create a very elongated ellipsoid.  Also added 'square' option.  

I'm working on a fix for the bcc tests (the rotation angles need to be converted to the new coordinate system).

RichardHeenan commented:
Hi Paul,
Thank you, will have a look in the morning ...
Richard

________________________________
From: Paul Kienzle [notifications@github.com]
Sent: 23 October 2017 17:48
To: SasView/sasmodels
Cc: Heenan, Richard (STFC,RAL,ISIS); State change
Subject: Re: [SasView/sasmodels] Ticket 776 orientation (#52)


Added size option to command line. For example, explore/jitter.py ellipsoid 40,40,4000 would create a very elongated ellipsoid. Also added 'square' option.

I'm working on a fix for the bcc tests (the rotation angles need to be converted to the new coordinate system).

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub<https://github.com/SasView/sasmodels/pull/52#issuecomment-338722651>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEB8pk13mNroqrsMdCoO7iOyiT2_3uwVks5svMNOgaJpZM4QCJup>.

pkienzle commented:
I just noticed that I've changed the sign of zhat, but xhat and yhat stayed the same.  I'm inclined to stay with the new definitions since the equations are auto-generated from angles/explore.py from the rotation matrices, which makes it much easier to explain what is going on.  Either that or update angles.py and jitter.py to use the 4.0 coordinate system.

pkienzle commented:
The change in sign in zhat explains the failing fcc and bcc tests.

RichardHeenan commented:
Have been rotating triaxial ellipsoid in sasview branch and in explore/jitter, which seem to agree with each other.

I think that we need to expand the users docs to explain that the theta rotation of the particle  is "done first" in the xz plane, then the phi rotation, as my current diagram is actually ambiguous. (The prototype docs are the ones for elliptical_cylinder, have a look at them, yes they need more work.) Paul K - how would you describe the order of the three rotations?

There is a growing consensus that to avoid confusing users the rotation angles should not be called theta, phi & psi, as they are also used with different meanings in the form factor equations. But what should we call them? Alpha, beta & gamma also have to be avoided.

pkienzle commented:
I resolved issue wherein check1d returned inconsistencies between 1D and 2D with dispersity.  

Unfortunately, we cannot use the same trick that the 1D models use of fixing the sample in the canonical orientation and scanning theta and phi all around.  Instead we need to include the rotation about psi.  Once that is included then most of the models approximately match for reasonable values of the parameters.  Closer matches are possible using double precision and increasing the number of points in the integration, but the numerical triple integral is expensive so I only tested this for a couple of cases.

Models with a very long dimension need more integration points.  In this case the 1D model is also very likely to be wrong since none of them modify the order of integration based on the shape dimensions.

fcc/bcc/sc are different at low q.  I already noted in a ticket that these models are incorrect in 1D at low Q, so I didn't investigate too closely.

core_shell_parallelepiped 1D does not match unless thick_rim_c=0.  Not surprising since the 1D model doesn't include thick_rim_c in its calculation.

pkienzle commented:
Once the bcc and fcc tests are removed (or perhaps the models themselves removed since they are incorrect at low q), then the pull request can be merged.

dehoni commented:
I have played around with jitter and a long cylinder (10,10,1000) with rectangular distribution. Setting the prefered orientation along the beam direction, you will notice that the distribution in the dtheta distribution angle will be weighted with cos(theta) as it should be for the polar angle integration. In contrast, the dphi distribution is not weighted. In this choosen case, dphi and dtheta should produce the same scattering pattern simply turned 90 deg to each other (for one, the particle wiggles in horizontal, for the other in the vertical direction).
Also a test with Sasview shows that this is not the case (see attached files):
[Circ avg cylinder2d dphi90.txt](https://github.com/SasView/sasmodels/files/1416508/Circ.avg.cylinder2d.dphi90.txt)
[Circ avg cylinder2d dtheta90.txt](https://github.com/SasView/sasmodels/files/1416509/Circ.avg.cylinder2d.dtheta90.txt). 
To get it correctly the distribution angles have to be reparametrized taking into account the component parallel to the deflection angle theta and the component perpendicular to theta, i.e. in spherical coordinates.

pkienzle commented:
In the current scheme theta and phi act differently, with delta theta as a segment of latitude, giving a circular cross section, and delta phi as a segment of longitudes, giving a section more akin to an orange slice.  You can see this in jitter by setting small dTheta-large dPhi  vs  large dTheta-small dPhi.  Notice in particular that the different phi points get closer together at large theta.  In order to make the numerical integration work correctly, the integration points need to be weighted by cos(theta).

I believe your proposed angle redefinition will have the equivalent problem, which is that with increasing theta, the phi points will get closer together.  Indeed, other than a change of axes (we have beam direction along -z), I think we are already implementing it.

So long as we are calculating the integral from a square mesh on a spherical surface we are going to have this problem.  We may be able to do better by defining a patch on the surface and sampling uniformly from it, but this will not fit nicely with the sasview's current support for dispersity distributions.

In lieu of a correct solution, we should recommend in the docs that phi should be the wider distribution and theta the narrower distribution, adjusting the choice of a, b and c axes for triaxial shapes to match.  

You can use sascomp to show the distortion by setting a comparison between phi=0, phi_pd=10, theta_pd=1 and phi=90, phi_pd=1, theta_pd=10:
```
sascomp -2d cylinder theta=0 phi=0,90 theta_pd_type=rectangle phi_pd_type=rectangle phi_pd=10,1 theta_pd=1,10 length=500 radius=10
```

pkienzle commented:
Visual artifacts are showing up for some choices of parameters on single precision:
```
sascomp -2d cylinder theta_pd_n=35 phi_pd_n=35 theta=0 phi=0 phi_pd=30 theta_pd=1 length=10000 radius=10 -single!
```

May want to try a quaternion calculation instead of rotation matrices since they are said to be more stable.

dehoni commented:
Did some systematic test with sascomp using
'''
python sascomp -2d cylinder theta=0 phi=0,90 theta_pd_type=rectangle phi_pd_type=rectangle phi_pd=1,n theta_pd=n,1 length=500 radius=10
'''
for several angular distribution width (n=90,70,50,30,10,5,2).
In principle, as far as I understand it, the produced scattering pattern for each distribution set should be more or less identical, since the particles have in both cases a angular spread along the vertical direction. For the first using dtheta alone, for the other the particle is rotated by 90 deg around the long axis and a dispersion on dphi .

The intensity dependence in angular spread clearly shows that with increasing width of the distribution, the results disagree
n=90 Intensity 1810154 1421764
n=70 Intensity 2056603 1513569
n=50 Intensity 2203200 1803208
n=30 Intensity 2594586 2441930
n=10 Intensity 4410125 4395022
n=5 Intensity 5497853 5496193
n=2 Intensity 6021116 6021078 .
I assume that it comes from the fact that the effective polar angle weight is applied for one distribution  but not for the other, which in the end rescales the intensity.

pkienzle commented:
I've pushed a new version of jitter with an alternate scheme: fix the arc length of the phi distribution independent of theta.   This keeps the points roughly equally spaced in both phi and theta, which should make dphi and dtheta equivalent when theta=0 

In order to make the total integral work out, I chop the phi points that wrap.  For phi=180, cos(theta)*n points will be kept, which will lead to approximately a cos(theta) weighting of the complete surface integral.

It is not clear to me that chopping is the correct behaviour for a gaussian distribution, but then the gaussian distribution is probably not the correct description of particle behaviour for large dispersity.

I will put some ifdefs in sasmodels to see if the full integral actually works in practice.

pkienzle commented:
kernel_iq.c now has a flag ```USE_SCALED_DPHI``` at the top which flips between wedges and rings for dphi steps.  The ring mode removes the distinction between dphi and dtheta in jitter, and by clipping phi in [-pi, pi], still allows the spherical integration checks to pass.

I now leave it to the theorists to decide which behaviour they prefer.

pkienzle commented:
As a further test, take a random parallelepiped at a random view and spin theta through 180 degrees; this should be identical to the same parallelepiped with a and b swapped, psi rotated by 90 and spin phi through 180 degrees.  This matches to within a couple of digits of machine precision:
```
sascomp -2d parallelepiped length_a=,length_b length_b=,length_a psi=0,90 theta_pd_n=100,0 phi_pd_n=0,100 theta_pd_type=rectangle phi_pd_type=rectangle theta_pd=103.92 phi_pd=103.92 -double -random
```

It is easy to imagine systems which can rotate freely in one dimension but not the other so being able to treat theta like phi and psi is preferable.

pkienzle commented:
I've been using lower case psi rather than uppercase Psi.

pkienzle commented on sasmodels/models/bcc_paracrystal.py:
```diff
@@ -64,19 +64,23 @@
-be accurate.
+be accurate, particularly at low $q$. For general details of the calculation and angular 
+dispersions for oriented particles see :ref:`orientation` .
+Note that we are not responsible for any incorrectness of the 2D model computation.
```
Rather than "we are not responsible for any incorrectness", how about specifying what we did to check the model, or perhaps "the 2D model equations have not been verified.".  The overall BSD license that Andrew threw on in April already disclaims any responsibility.

pkienzle commented:
The problem of mapping a mesh onto the sphere is equivalent to a map projection.

There are a variety of these given in wikipedia:

    https://en.wikipedia.org/wiki/List_of_map_projections

Obviously none of them will be perfect, but we are looking for one that treats theta and phi equivalently without distorting the grid of theta-phi points too much.  The following is particularly promising, since it preserves distance on the surface:

    https://en.wikipedia.org/wiki/Azimuthal_equidistant_projection

I've implemented it in jitter so you can look at the point distribution.  Basically, I overlay a square mesh of +/- 180 degrees in phi and theta, chopping all points that lie outside the disk.  Still haven't figured out how to weight the points so that the integrals will work out.

pkienzle commented:
An alternative approach is to scrap the mesh for orientation, and go directly to distributions on the sphere.  This will be a more radical change, pushing the interpretation of theta-phi distributions into the kernel code, and fixing them either as uniform or as some form of gaussian.

The Kent distribution (https://en.wikipedia.org/wiki/Kent_distribution) claims to be a gaussian-like distribution on the surface of a sphere.  I have a partial implementation of this in jitter.py, but couldn't make it produce reasonable values for probability.  The azimuthal equidistance projection is sufficiently promising, that we can hold off on more radical alternatives for now.

pkienzle commented:
New map projection 'guyou' added to jitter.py.  See docs for draw_mesh at jitter.py:172.  Play with it using:
```
    explore/guyou.py
    explore/jitter.py -p guyou
```
It seems to meet the desired criteria of acting similarly in theta and phi, and supporting free rotation in either with linear spacing between the points, with jitter almost as uniform as jitter for small theta, uniform phi in the simple polar coordinates model.

Issues: 
(1) haven't yet computed the weighting function 
(2) haven't yet implemented in opencl
(2) it will be considerably more expensive to compute than the simple polar coordinates model, though maybe a lookup table would be good enough

pkienzle commented:
Need to decide how to move forward on this ticket.  Brief summary:

* equirectangular (the standard form) is good enough, with the caveat that phi width should be greater than theta width, and that axes should be relabelled to make it so (only needed if there is substantial orientation dispersity in phi)

* azimuthal_equidistance provides excellent support for small deviations; being distance preserving, gaussian ellipses should stay as gaussian ellipse.

* sinusoidal (scaled phi) is a relatively simple mechanism for treating theta and phi independently, but it behaves badly at theta +/- 90

* guyou is an excellent compromise, introducing a little distortion but otherwise treating theta and phi almost identically. However, it is not yet complete, and it may be expensive.

My recommendation is to go with the good enough and fast simple representation for now, and put guyou support in as a ticket for later.

RichardHeenan commented:
Need to decide how to move forward on this ticket. Brief summary:
> some comments from Richard

    equirectangular (the standard form) is good enough, with the caveat that phi width should be greater than theta width, and that axes should be relabelled to make it so (only needed if there is substantial orientation dispersity in phi)
> need to change the user instructions in sasmodels/doc/guide/orientation/orientation.rst to reflect this
As I understand it, jitter in theta and phi are about a & b axes of particle (which I labelled delta1 and delta2 in the diagram to make the point - but should we change this?). Thus far we have said keep a<b<c, but  we have to tell users that that may need to have b<a<c if that better suits the jitter.  But does swapping a & b have issues in any of the asymmetric particle models?  I would also add some detail about the limitations of the rectangular grid for jitter integration.

    azimuthal_equidistance provides excellent support for small deviations; being distance preserving, gaussian ellipses should stay as gaussian ellipse.


    sinusoidal (scaled phi) is a relatively simple mechanism for treating theta and phi independently, but it behaves badly at theta +/- 90

    guyou is an excellent compromise, introducing a little distortion but otherwise treating theta and phi almost identically. However, it is not yet complete, and it may be expensive.

My recommendation is to go with the good enough and fast simple representation for now, and put guyou support in as a ticket for later.
>  I agree. We understand the issues of the rectangular grid much better now and the new jitter in particle coordinates is much better than current system. Just need to make sure users are adequately informed so they don't push things too far. Merging all this soon will help iron out bugs.

RichardHeenan commented:
Yet more to sort out ...

At least in my latest build of this branch, **triaxial_ellipsoid** and **parallelepiped** have the _a_ axis horizontal when _theta=phi=psi=0_ as I expect, but **elliptical_cylinder** and **core_shell_bicelle_elliptical** have it vertical (looking at the Q dependence of a compute with 2d data). Can someone else please check and then help decide how to fix this.

Also **hollow_cylinder** refuses to compute or fit 2d data, though theta & phi boxes appear and it passes a 2d data unit test.

pkienzle commented:
Confirmed that elliptical_cylinder is rotated relative to docs---Iqxy should use r_ratio*qb rather than r_ratio*qa.  Note the ```phi=0,90``` required to get parallelepiped and elliptical_cylinder to have a similar repeat structure:

```
sascomp elliptical_cylinder,parallelepiped theta=0 phi=0 psi=0 -2d -midq radius_minor=17.5 axis_ratio=2 length_b=70 phi=0,90
```

Similarly, core_shell_bicelle_elliptical should use r_major*qb, r_minor*qa, etc., which you can see because it aligns with a similar elliptical_cylinder with no phi rotation:

```
sascomp elliptical_cylinder,core_shell_bicelle_elliptical theta=0 phi=0 psi=0 -pars -2d -highq radius_minor=17.5 axis_ratio=2 radius=15 x_core=2 thick_rim=2.5 length=400,350 thick_face=25  sld_rim=sld sld_face=sld sld_core=sld -highq
```

These should change to be consistent with the documentation.

Probably true of the new belt_rough model, but didn't check.

pkienzle commented:
The hollow cylinder problem showed up a subtle bug that would have been quite nasty otherwise: On occasion, depending on which polydispersity parameters were in effect, the jitter distribution may have been set to the distribution (value=[view_angle], weight=[1]).

Most models appear to work now, except:

* rpa is returning zero
* spherical_sld is not even displaying the form

Both of these are problems with master, not with this branch.

gonzalezma commented:
I guess the problem with the spherical_sld is related to the problem reported by Paul B (ticket 1007). I looked into this and have a branch in sasview where the problem is fixed for this model. The error comes from self.model.fun_list being a tuple, but called like a dictionary in basepage.py. In the branch, I corrected this for the spherical_sld model, but as I don't know which are the other situations in which fun_list is used, I did not go beyond that and thus the job is not finished. If someone tells me which are the other cases to check, I could try to complete it.

RichardHeenan commented:
Hi Paul K

I could swap the a & b axes of elliptical_cylinder etc, as I presume that you have not done this.

Shall I do this in sasmodels/ticket-776-orientation ? though I see you merged this into sasmodels/ticket-786

Could also do with having the $ math change in my docs builds, from your sasview/docs branch. You have asked Steve King to review, but he is not back from holidays until 13 or 14 Nov. Perhaps someone else could look at this?

More generally to anyone reading this, from where can I obtain the latest set up instructions and yml script for the new tk gui ? What I did briefly have working on my laptop at code camp no longer wants to work.

Thanks

Richard

From: Paul Kienzle [mailto:notifications@github.com]
Sent: 02 November 2017 18:56
To: SasView/sasmodels
Cc: Heenan, Richard (STFC,RAL,ISIS); State change
Subject: Re: [SasView/sasmodels] Ticket 776 orientation (#52)


Confirmed that elliptical_cylinder is rotated relative to docs---Iqxy should use r_ratioqb rather than r_ratioqa. Note the phi=0,90 required to get parallelepiped and elliptical_cylinder to have a similar repeat structure:

sascomp elliptical_cylinder,parallelepiped theta=0 phi=0 psi=0 -2d -midq radius_minor=17.5 axis_ratio=2 length_b=70 phi=0,90

Similarly, core_shell_bicelle_elliptical should use r_majorqb, r_minorqa, etc., which you can see because it aligns with a similar elliptical_cylinder with no phi rotation:

sascomp elliptical_cylinder,core_shell_bicelle_elliptical theta=0 phi=0 psi=0 -pars -2d -highq radius_minor=17.5 axis_ratio=2 radius=15 x_core=2 thick_rim=2.5 length=400,350 thick_face=25  sld_rim=sld sld_face=sld sld_core=sld -highq

These should change to be consistent with the documentation.

Probably true of the new belt_rough model, but didn't check.

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub<https://github.com/SasView/sasmodels/pull/52#issuecomment-341524046>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEB8puwfoV6cNqeb0s7U4eadO03FnfyTks5syhBRgaJpZM4QCJup>.

pkienzle commented:
*I could swap the a & b axes of elliptical_cylinder etc, as I presume that you have not done this. Shall I do this in sasmodels/ticket-776-orientation ? though I see you merged this into sasmodels/ticket-786*

Yes, please update them.  You could do so in a new branch based off of ticket-776-orientation like I did for core_shell_parallelepiped (which is inconvenient for us but more convenient for reviewers) or you can do so in ticket-776-orientation.

As far as I know, the only thing holding up this branch now are the test values for bcc/fcc_paracrystal, which did change when I made the 2D calc consistent with the 1D calc.
